services/storehost/extStats.go

```diff
@@ -0,0 +1,434 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package storehost
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pborman/uuid"
+	"github.com/uber-common/bark"
+	"github.com/uber/cherami-server/common"
+	"github.com/uber/cherami-thrift/.generated/go/metadata"
+	"github.com/uber/cherami-thrift/.generated/go/shared"
+)
+
+const (
+	reportChanBufLen      = 1024
+	defaultReportInterval = time.Duration(time.Minute)
+)
+
+type (
+	ExtStatsReporter struct {
+		hostID  string
+		xMgr    *ExtentManager
+		mClient metadata.TChanMetadataService
+		logger  bark.Logger
+
+		sync.RWMutex
+		extents    map[string]*extStatsContext
+		reportPool sync.Pool
+		reportC    chan *report
+
+		wg    sync.WaitGroup
+		stopC chan struct{} // internal channel to stop
+	}
+
+	extStatsContext struct {
+		ext        *extentContext
+		ref        int64
+		lastReport *report
+	}
+
+	report struct {
+		ctx       *extStatsContext
+		extentID  uuid.UUID
+		timestamp int64
+
+		firstSeqNum, lastSeqNum int64
+		sealed                  bool
+	}
+)
+
+// the following globals and their methods are
+var (
+	reportInterval         atomic.Value
+	extStatsReporterPaused int32 = 0
+)
+
+func init() {
+	reportInterval.Store(defaultReportInterval)
+}
+
+// ExtStatsReporterPause pauses the reporting (intended for tests)
+func ExtStatsReporterPause() {
+
+	swapped := atomic.CompareAndSwapInt32(&extStatsReporterPaused, 0, 1)
+	common.GetDefaultLogger().WithField(`already-paused`, !swapped).Info("extStatsReporter: paused")
+}
+
+// ExtStatsReporterUnpause unpauses the reporting (intended for tests)
+func ExtStatsReporterUnpause() {
```

Comment:
done. i left it as 'unpause' because that's the name used by 'queueMonitor" (that this supplants) and in the integration_test that calls into this. nevertheless, i made this 'resume' and changed the call within integration_test too.